xarray/core/indexing.py

```diff
@@ -654,6 +692,204 @@ def _outer_to_numpy_indexer(key, shape):
+    np_indexer = []
+    # posify
+    indexer = [np.where(k < 0, k + s, k) if isinstance(k, np.ndarray) else k
+               for k, s in zip(indexer.tuple, shape)]
+
+    if indexing_support is IndexingSupport.OUTER_1VECTOR:
+        # some backends such as h5py supports only 1 vector in indexers
+        # We choose the most efficient axis
+        gains = [(np.max(k) - np.min(k) + 1.0) / len(np.unique(k))
+                 if isinstance(k, np.ndarray) else 0 for k in indexer]
+        array_index = np.argmax(np.array(gains)) if len(gains) > 0 else None
+
+        for i, k in enumerate(indexer):
+            if isinstance(k, np.ndarray) and i != array_index:
+                backend_indexer.append(slice(np.min(k), np.max(k) + 1))
+                np_indexer.append(k - np.min(k))
+            elif isinstance(k, np.ndarray):
+                pkey, ekey = np.unique(k, return_inverse=True)
+                backend_indexer.append(pkey)
+                np_indexer.append(ekey)
+            else:
+                backend_indexer.append(k)
+                np_indexer.append(slice(None))
+
+        return (OuterIndexer(tuple(backend_indexer)),
+                (OuterIndexer(tuple(np_indexer)), ))
+
+    if indexing_support == IndexingSupport.OUTER:
+        for k in indexer:
+            if (isinstance(k, integer_types + (slice, )) or
+                    (np.diff(k) >= 0).all()):
+                backend_indexer.append(k)
+                np_indexer.append(slice(None))
+            else:  # not sorted np.ndarray indexer
+                oind, vind = np.unique(k, return_inverse=True)
+                backend_indexer.append(oind)
+                np_indexer.append(vind.reshape(*k.shape))
+
+        return (OuterIndexer(tuple(backend_indexer)),
+                (OuterIndexer(tuple(np_indexer)), ))
+
+    # basic
+    for i, k in enumerate(indexer):
+        if isinstance(k, np.ndarray):
+            backend_indexer.append(slice(np.min(k), np.max(k) + 1))
+            np_indexer.append(k - np.min(k))
+        else:
+            backend_indexer.append(k)
+            np_indexer.append(slice(None))
+
+    return (BasicIndexer(tuple(backend_indexer)),
+            (OuterIndexer(tuple(np_indexer)), ))
+
+
+def _arrayize_vectorized_indexer(indexer, shape):
+    """ Return an identical vindex but slices are replaced by arrays """
+    slices = [v for v in indexer.tuple if isinstance(v, slice)]
+    if len(slices) == 0:
+        return indexer
+
+    arrays = [v for v in indexer.tuple if isinstance(v, np.ndarray)]
+    n_dim = arrays[0].ndim if len(arrays) > 0 else 0
+    i_dim = 0
+    new_key = []
+    for v, size in zip(indexer.tuple, shape):
+        if isinstance(v, np.ndarray):
+            new_key.append(np.reshape(v, v.shape + (1, ) * len(slices)))
+        else:  # slice
+            shape = ((1,) * (n_dim + i_dim) + (-1,) +
+                     (1,) * (len(slices) - i_dim - 1))
+            new_key.append(np.arange(*v.indices(size)).reshape(shape))
```

Comment:
According to the tests, `v` is sometimes an integer at this point, and thus has an attribute error for `v.indices`.
From the test `TestLazyArray.test_vectorized_lazily_indexed_array`:

```
indexer = BasicIndexer((slice(1, 3, 2), 0)), shape = (1, -1)
    def _arrayize_vectorized_indexer(indexer, shape):
        """ Return an identical vindex but slices are replaced by arrays """
        slices = [v for v in indexer.tuple if isinstance(v, slice)]
        if len(slices) == 0:
            return indexer
    
        arrays = [v for v in indexer.tuple if isinstance(v, np.ndarray)]
        n_dim = arrays[0].ndim if len(arrays) > 0 else 0
        i_dim = 0
        new_key = []
        for v, size in zip(indexer.tuple, shape):
            if isinstance(v, np.ndarray):
                new_key.append(np.reshape(v, v.shape + (1, ) * len(slices)))
            else:  # slice
                shape = ((1,) * (n_dim + i_dim) + (-1,) +
                         (1,) * (len(slices) - i_dim - 1))
>               new_key.append(np.arange(*v.indices(size)).reshape(shape))
E               AttributeError: 'int' object has no attribute 'indices'
```

Reply:
Thanks. 
Our VectorizedIndexer only consists of np.ndarrays and slices, it is OK here.
Indeed, the error was due to another bug. 
Fixed.